src/Layers/xrRenderDX10/dx10HWCaps.cpp

```diff
@@ -0,0 +1,171 @@
+#include "stdafx.h"
+#pragma hdrstop
+
+#include "Layers/xrRender/HWCaps.h"
+#include "dx10HW.h"
+
+#if !defined(_EDITOR)
+#include <nvapi.h>
+#include <ags_lib/inc/amd_ags.h>
+#endif
+
+namespace
+{
+#if !defined(_EDITOR)
+u32 GetNVGpuNum()
+{
+    NvLogicalGpuHandle logicalGPUs[NVAPI_MAX_LOGICAL_GPUS];
+    NvU32 logicalGPUCount;
+    NvPhysicalGpuHandle physicalGPUs[NVAPI_MAX_PHYSICAL_GPUS];
+    NvU32 physicalGPUCount;
+
+    //	int result = NVAPI_OK;
+
+    int iGpuNum = 0;
+
+    NvAPI_Status status;
+    status = NvAPI_Initialize();
+
+    if (status != NVAPI_OK)
+    {
+        Msg("* NVAPI is missing.");
+        return iGpuNum;
+    }
+
+    // enumerate logical gpus
+    status = NvAPI_EnumLogicalGPUs(logicalGPUs, &logicalGPUCount);
+    if (status != NVAPI_OK)
+    {
+        Msg("* NvAPI_EnumLogicalGPUs failed!");
+        return iGpuNum;
+        // error
+    }
+
+    // enumerate physical gpus
+    status = NvAPI_EnumPhysicalGPUs(physicalGPUs, &physicalGPUCount);
+    if (status != NVAPI_OK)
+    {
+        Msg("* NvAPI_EnumPhysicalGPUs failed!");
+        return iGpuNum;
+        // error
+    }
+
+    Msg("* NVidia MGPU: Logical(%d), Physical(%d)", physicalGPUCount, logicalGPUCount);
+
+    //	Assume that we are running on logical GPU with most physical GPUs connected.
+    for (u32 i = 0; i < logicalGPUCount; ++i)
+    {
+        status = NvAPI_GetPhysicalGPUsFromLogicalGPU(logicalGPUs[i], physicalGPUs, &physicalGPUCount);
+        if (status == NVAPI_OK)
+            iGpuNum = _max(iGpuNum, physicalGPUCount);
+    }
+
+    if (iGpuNum > 1)
+        Msg("* NVidia MGPU: %d-Way SLI detected.", iGpuNum);
+
+    return iGpuNum;
+}
+
+u32 GetATIGpuNum()
+{
+    AGSContext* ags = nullptr;
+    AGSGPUInfo gpuInfo = {};
+    AGSReturnCode status = agsInit(&ags, &gpuInfo);
+    if (status != AGS_SUCCESS)
+    {
+        Msg("* AGS: Initialization failed (%d)", status);
+        return 0;
+    }
+    int crossfireGpuCount = 1;
+    status = agsGetCrossfireGPUCount(ags, &crossfireGpuCount);
+    if (status != AGS_SUCCESS)
+    {
+        Msg("! AGS: Unable to get CrossFire GPU count (%d)", status);
+        agsDeInit(ags);
+        return 1;
+    }
+    Msg("* AGS: CrossFire GPU count: %d", crossfireGpuCount);
+    agsDeInit(ags);
+    return crossfireGpuCount;
+}
+
+u32 GetGpuNum()
+{
+    u32 res = GetNVGpuNum();
+    res = _max(res, GetATIGpuNum());
+    res = _min(res, CHWCaps::MAX_GPUS);
+
+    if (res == 0)
+    {
+        Log("! Cannot find graphic adapter. Assuming that you have one...");
+        res = 1;
+    }
+
+    Msg("* Starting rendering as %d-GPU.", res);
+
+    return res;
+}
+#else
+u32 GetGpuNum() { return 1; }
+#endif
+}
+
+
```

Comment:
Please, remove this one empty lines. Two empty lines is excessive, one is enough.